Sources/PackagePlugin/Implementation.swift

```diff
@@ -0,0 +1,132 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+@_implementationOnly import Foundation
+
+// The way in which SwiftPM communicates with the package plugin is an im-
+// plementation detail, but the way it currently works is that the plugin
+// is compiled (in a very similar way to the package manifest) and then run in
+// a sandbox. Currently it is passed the JSON encoded input structure as the
+// last command line argument; however, it this will likely change to instead
+// read it from stdin, since that avoids any command line length limitation.
+// Any generated commands and diagnostics are emitted on stdout after a zero
+// byte; this allows regular output, such as print statements for debugging,
+// to be emitted to SwiftPM verbatim. SwiftPM tries to interpret any stdout
+// contents after the last zero byte as a JSON encoded output struct in UTF-8
+// encoding; any failure to decode it is considered a protocol failure. The
+// exit code of the compiled plugin determines success or failure (though
+// failure to decode the output is also considered a failure to run the ex-
+// tension).
+
+extension Plugin {
+
+    public static func main() throws {
```

Comment:
should we support async main too?

Reply:
At the moment that would mean that plugins would require macOS 12, I think.  This came up in the pitch; I'll check what the status of the system requirements are.

Reply:
This has come up again in the proposal for the amendment.  If we don't think that the back deployment problem will be a concern, then this does seem like the right, forward-looking thing to do.